File: examples/js/src/tools/executeCommand.ts
import { tool } from "@langchain/core/tools";
import { execaCommand, ExecaSyncError } from "execa";

const executeCommand = tool(
  async ({ command }) => {
    console.log(`Executing command: ${command}`);
    try {
      const { all } = await execaCommand(command, { shell: true, all: true });

      let result = all.trim();

      if (result.length > 10000) {
        result = result.slice(0, 10000) + "\n\n...truncated...";
      }

      //console.log(`Command output: ${result}`);

      return result;
    } catch (error) {
      if (error instanceof ExecaSyncError) {
        console.error(`Command failed: ${command}`);
        const errorMessage = `Command failed with exit code ${error.exitCode}: ${command}\n\nError: ${error.stderr || error.message}\n\nTip: If using git commands with options and file paths, make sure options (like --stat) come before file paths (like index.mjs).`;
        console.log(`Command error: ${errorMessage}`);
        return errorMessage;
      }

      throw error;
    }
  },
  {
    name: "execute",
    description: "Execute a command line command",
    schema: {
      type: "object",
      properties: {
        command: { type: "string", description: "Command to execute" },
      },
      required: ["command"],
      additionalProperties: false,
    },
  },
);

export default executeCommand;

File: examples/python/src/tools/execute_command.py
import subprocess
from langchain_core.tools import tool


@tool
def execute(command: str) -> str:
    """Execute a command line command"""
    print(f"Executing command: {command}")
    try:
        result = subprocess.run(
            command, shell=True, capture_output=True, text=True
        )
        output = (result.stdout + result.stderr).strip()

        if len(output) > 10000:
            output = output[:10000] + "\n\n...truncated..."

        return output
    except Exception as error:
        error_message = (
            f"Command failed: {command}\n\n"
            f"Error: {error}\n\n"
            f"Tip: If using git commands with options and file paths, "
            f"make sure options (like --stat) come before file paths (like index.mjs)."
        )
        print(f"Command error: {error_message}")
        return error_message
